### PR TITLE
Prometheus Push Gateway for Stickerization Request Metrics

### PR DESCRIPTION
**Pull request by derhnyel**
This creates prometheus gauges for stickerization metrics ,gets the metrics, increments the guages and pushes to an enpoint . This solves half of  #20 . @deven96  @MeNsaaH , You can now work on the prometheus and grafana end of things

### CODE REVIEW

**Review by MeNsaaH — CHANGES_REQUESTED** [review]
Can you add an architectural diagram of the current setup. Detailing the master, worker, redis and prometheus gateway setup for proper visualization?

File: docker-compose.yml [hunk 1]
```diff
@@ -37,10 +38,30 @@ services:
...
+    command: ["--persistence.file","/tmp/whatsticker.metrics"]
+    restart: always
+    ports: 
+      - "9091:9091" 
+    expose:
+      - "9091"
+    volumes:
+      - type: bind
+        source: ./whatsticker.metrics
+        target: /tmp/whatsticker.metrics
```
Comment (deven96) [inline docker-compose.yml:64]: were you able to get the metrics to be bound to the persistence file?

Comment (deven96) [reply docker-compose.yml:64]: @derhnyel tell Onuh he has to add a `whatsticker.metrics` file to the repo in prod before redeploying

File: logger/main.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	<-c
+}
+
+func main() {
+
+	gauges := metrics.NewGauges()
+	registry := metrics.NewRegistry()
+	metric := metrics.Initialize(registry, gauges)
+
+	log.Printf("Initialized Metrics Consumer %#v", metric)
```
Comment (deven96) [inline logger/main.go:26]: Nit: Change to `Initialized Metrics SideCar`

**Comment by derhnyel** [comment]
> Can you add an architectural diagram of the current setup. Detailing the master, worker, redis and prometheus gateway setup for proper visualization?

Okay, will do that in the readme.

File: logger/metrics/metrics.go [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+
+	if stickerMetric.MediaType == "image" {
+		stickerGauges.ImageGauge.Inc()
+	} else if stickerMetric.MediaType == "video" {
+		stickerGauges.VideoGauge.Inc()
+	} else {
+		stickerGauges.InvalidMediaGauge.Inc()
+	}
+
+	if stickerMetric.Validation {
```
Comment (deven96) [inline logger/metrics/metrics.go:161]: Nit: Change to `Validated` instead of `Validation`

File: logger/metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+
+func extractCountry(number string) string {
+	phoneNumber := strings.Trim(number, "+")
+	country := phonenumber.GetISO3166ByNumber(phoneNumber, true)
+	fmt.Println(country.CountryName)
+	return country.CountryName
+
+}
+
+func (consumer *MetricConsumer) Consume(delivery rmq.Delivery) {
```
Comment (deven96) [inline logger/metrics/metrics.go:176]: Nice work on the logger in general... you grabbed the concept quite well

File: master/Dockerfile [hunk 5]
```diff
@@ -9,3 +9,6 @@ ADD handler ./handler
 ADD task ./task
 RUN go mod tidy
 ENTRYPOINT ["go", "run", "main.go", "-log-level", "DEBUG"]
+
```
Comment (deven96) [inline master/Dockerfile:12]: Get rid of the whitespaces here

File: master/handler/handler.go [hunk 6]
```diff
@@ -32,13 +34,28 @@ type Handler interface {
 	// also sends message to client about issue
 	Validate() error
 	// Handle : obtains the message to be sent as response
-	Handle(pushTo rmq.Queue)
+	Handle(pushTo rmq.Queue, PushMetricsTo rmq.Queue)
```
Comment (deven96) [inline master/handler/handler.go:37]: Instead of passing it to `Handle` pass it to `SetUp` and use it later

File: worker/main.go [hunk 7]
```diff
@@ -28,10 +28,13 @@ func main() {
 	}
 	completeQueue, _ := connection.OpenQueue(os.Getenv("SEND_TO_WHATSAPP_QUEUE"))
 	convertQueue, _ := connection.OpenQueue(os.Getenv("CONVERT_TO_WEBP_QUEUE"))
+
 	convert := &convert.ConvertConsumer{
 		// set to push to completeQueue when done
+		//set to push metrics to loggingQueue when done
```
Comment (deven96) [inline worker/main.go:34]: Comment is now stale as worker doesn't push metrics anymore no?

File: worker/convert/convert.go [hunk 8]
```diff
@@ -66,21 +68,24 @@ func (consumer *ConvertConsumer) Consume(delivery rmq.Delivery) {
...
 	}
-
 	err := cmd.Run()
 	if err != nil {
 		fmt.Printf("Failed to Convert %s to WebP %s", task.MediaType, err)
 		return
 	}
 
+	//Function to Get file lenght
+
```
Comment (deven96) [inline worker/convert/convert.go:78]: nit: Remove unnecessary whitespace in file

**Review by deven96 — COMMENTED** [review]
I keep thinking about it more and more and I believe that the `logging` and `metrics` service can be condensed into one if the logging service were to serve http traffic on a port instead of having to push it to metrics again. 


We might no longer have use for metrics since you solved the logging issue by creating a logging service. All we need to do is setup a http endpoint using their own prometheus examples

**Review by deven96 — CHANGES_REQUESTED** [review]
Merge metrics and logging into one happy family

File: docker-compose.yml (hunk 1 above)
Comment (derhnyel) [reply docker-compose.yml:64]: Not yet , i can't `chown` on my current OS. I found a solution here [:](https://stackoverflow.com/questions/50009065/how-to-persist-data-in-prometheus-running-in-a-docker-container)

Comment (derhnyel) [reply docker-compose.yml:64]: Yeah.. I will .. Also i still need to make 1 more  commit for this to be merge ready

File: logger/metrics/metrics.go (hunk 3 above)
Comment (derhnyel) [reply logger/metrics/metrics.go:161]: Okay... will doo

File: worker/main.go (hunk 7 above)
Comment (derhnyel) [reply worker/main.go:34]: Exactly

File: master/handler/handler.go (hunk 6 above)
Comment (derhnyel) [reply master/handler/handler.go:37]: I thought of that but it might not work.. because the errors we are trying to catch happen in the interface's handle method... so i have to pass the metrics queue to that method...  An Error such as 
```suggestion
	if err != nil {
		fmt.Printf("Failed to download image: %v\n", err)
		pushMetricsTo.PublishBytes(metricBytes)
		return
	}
```

**Review by deven96 — APPROVED** [review]
